core-util/SharedPointer.h

```diff
@@ -0,0 +1,275 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __SHAREDPOINTER_H__
+#define __SHAREDPOINTER_H__
+
+#include "core-util/core-util.h"
+#include "mbed-alloc/ualloc.h"
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifndef NDEBUG
+#include <stdio.h>
+#define SHAREDPOINTER_DEBUG(...) { printf(__VA_ARGS__); }
+#else
+#define SHAREDPOINTER_DEBUG(...) /* nothing */
+#endif
+
+/*
+    Shared pointer class.
+
+    Similar to std::shared_ptr in C++11.
+
+    Usage: SharedPointer<class> POINTER(new class())
+
+    When POINTER is passed around by value the copy constructor and
+    destructor counts the number of references to the original object.
+    If the counter reaches zero, delete is called on the object pointed to.
+
+    To avoid loops, "weak" references should be used by calling the original
+    pointer directly through POINTER.get().
+*/
+
+template <class T>
+class SharedPointer
+{
+public:
+    /*
+        Create empty SharedPointer not pointing to anything.
+        Used for variable declaration.
+    */
+    SharedPointer()
+        :   pointer(NULL),
+            counter(NULL)
+    {
+        SHAREDPOINTER_DEBUG("SP: %p [%p: %p]\r\n", this, pointer, counter);
+    }
+
+    /*
+        Create new SharedPointer and take control of object with address _pointer.
+    */
+    SharedPointer(T* _pointer)
+        :   pointer(_pointer)
+    {
+        CORE_UTIL_ASSERT(pointer);
+
+        // allocate counter on the heap so it can be shared
+        UAllocTraits_t traits = {0};
+        counter = (uint32_t*) mbed_ualloc(sizeof(uint32_t), traits);
+
+        // initialize counter to 1
+        CORE_UTIL_ASSERT(counter);
+        *counter = 1;
+
+        SHAREDPOINTER_DEBUG("SP: %p [%p: %p = %lu]\r\n", this, pointer, counter, *counter);
+    }
+
+    ~SharedPointer()
+    {
+        decrementCounter();
+    }
+
+    /*
+        Create new SharePointer from other SharedPointer by
+        copying pointer to original object and pointer to counter.
+    */
+    SharedPointer(const SharedPointer& source)
+        :   pointer(source.pointer),
+            counter(source.counter)
+    {
+        // increment reference counter
+        if (counter)
+        {
+            (*counter)++;
+        }
+
+        SHAREDPOINTER_DEBUG("SP&: %p = %p [%p: %p = %lu]\r\n", this, &source, pointer, counter, *counter);
+    }
+
+    /*
+        Assignment operator.
+
+        Cleanup previous reference and assign new pointer and counter.
+    */
+    SharedPointer operator=(const SharedPointer& source)
+    {
+        // clean up by decrementing counter
+        decrementCounter();
```

Comment:
If I'm reading this right, this will fail:

```
int *p = malloc(sizeof(int));
SharedPointer<int> p1(p);
SharedPointer<int> p2(p);
p1 = p2; // the assignment operator will call decrementCounter, which in turn will delete 'p' (since counter is 1), then it will assign the deleted pointer to itself
```

Does this make sense or I'm missing something?